core/arch/arm/pta/sdp_pta.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <kernel/pseudo_ta.h>
+#include <mm/core_memprot.h>
+#include <mm/tee_mmu.h>
+#include <sdp_pa.h>
+
+#define PTA_NAME "sdp.pta"
+
+static TEE_Result sdp_pa_cmd_virt_to_phys(uint32_t types,
+					  TEE_Param params[TEE_NUM_PARAMS])
+{
+	char *va = params[0].memref.buffer;
+	size_t len = params[0].memref.size;
+	paddr_t pa;
+	struct tee_ta_session *s;
+	struct user_ta_ctx *utc;
+	TEE_Result res;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_VALUE_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: %x", (unsigned int)types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	s = tee_ta_get_calling_session();
+	if (!s)
+		return TEE_ERROR_ACCESS_DENIED;
+
+	utc = to_user_ta_ctx(s->ctx);
+	res = tee_mmu_check_access_rights(utc, TEE_MEMORY_ACCESS_READ |
+					  TEE_MEMORY_ACCESS_WRITE |
+					  TEE_MEMORY_ACCESS_ANY_OWNER,
+					  (uaddr_t)va, len);
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_ACCESS_DENIED;
+
+	if (!core_vbuf_is(CORE_MEM_SDP_MEM, va, len)) {
+		DMSG("bad memref secure attribute");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	pa = virt_to_phys(va);
+
+	params[1].value.a = (uint32_t)pa;
+#if __SIZEOF_PADDR__ == 8
```

Comment:
You could use `reg_pair_from_64(pa, &params[1].value.b, &params[1].value.a)` instead.
Also I wonder why you put the most significant bits in `value.b`, wouldn't it make more sense to do the opposite?

Reply:
reg_pair_from_64() will be better. And I also will put most significant bits in value.a. The changing will be reg_pair_from_64(pa, &params[1].value.a, &params[1].value.b).